### PR TITLE
Widget Store update

### PR DESCRIPTION
**Pull request by lanice**
This revamps the frontend of the widget store, adding two side-panels:
1. Shop internals. Shows a table for the current inventory, as well as a table for all recorded orders, if existing. Orders are updated in real-time.
2. Server Tools. The "Crash the Application" button, with some explanatiory text.

Some backend changes were made to accommodate this, e.g., adding the scheduled workflow that "does work" on each order to bring it from PAID to DISPATCHED.

Unfortunately, due to some [issues](https://dbosinc.slack.com/archives/C05CDE2GR2P/p1724186595181399) of serving static files (and referencing them among themselves), all frontend code is in one big `app.html`. This is not ideal, but allows us to "just serve" the frontend via DBOS. In a real-world project, one would probably separate the frontend into its own app that could be deployed independently. In order to make this a single-deploy DBOS demo app, the current approach was chosen.

To play around, the app is deployed here: https://leanderneiss-widget-store.cloud.dbos.dev/

### CODE REVIEW

File: widget-store/public/app.html [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+            </template>
+          </div>
+        </div>
+
+        <div :class="visiblePanelOnMobile === 'tools' || 'hidden md:block'">
+          <div class="flex flex-col gap-6 p-5 text-center">
+            <h2 class="text-3xl text-center">Server Tools</h2>
+            <div x-html="divider"></div>
+            <p class="text-left text-gray-500">
+              Crash the application at any time to simulate a server failure.
```
Comment (kraftp) [inline widget-store/public/app.html:526]: I'd condense both paragraphs to something much shorter like this:

Crash the application at any time to simulate a server failure or service interruption. After crashing, the application will momentarily become inaccessible but then resume from exactly where it left off before the crash. Try it and see!

File: widget-store/public/app.html [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+              become inaccessible but then recover to its appropriate state,
+              allowing you to continue where you left off.
+              <br />
+              Any orders that have been paid and are still being proceessed
+              until they can be dispatched will continue to be processed once
+              the application is restarted.
+            </p>
+            <div>
+              <button
+                class="bg-red-600 border-none text-white py-4 px-6 text-center text-xl cursor-pointer rounded-lg transition duration-300 hover:bg-red-700 hover:shadow-lg"
```
Comment (kraftp) [inline widget-store/public/app.html:537]: Also, the tiny bomb doesn't look great--let's get rid of it.

File: widget-store/README.md [hunk 3]
```diff
@@ -19,7 +19,7 @@ To reset the inventory counter, run `npx dbos-cloud app deploy` again.
...
+Visit the main app page for an opportunity to buy a widget! Buying a widget decrements the remaining inventory. Clicking "Buy Now" takes you to a confirmation page that simulates payment processing. The Shop Internals section shows the current inventory and order status of all orders. The Server Tools section give you an opportunity to crash the app. After crashing, any in-progressAfter crashing, any in-progress payment page will momentarily become inaccessible but then recover to its appropriate state, allowing you to continue where you left off. After payment has been confirmed or denied for a specific worklow, revisiting `/?payment=old-ID` will not change the settled order.
```
Comment (qianl15) [inline widget-store/README.md:22]: Mention that the order status is updated every second?

File: widget-store/migrations/001_initial_schema.js [hunk 4]
```diff
@@ -13,6 +13,7 @@ exports.up = async function(knex) {
     table.datetime('last_update_time').notNullable().defaultTo(knex.fn.now());
     table.integer('product_id').notNullable();
     table.foreign('product_id').references('products.product_id');
+    table.integer('progress_remaining').notNullable().defaultTo(10);
```
Comment (qianl15) [inline widget-store/migrations/001_initial_schema.js:16]: Mention this new functionality somewhere, maybe both in the README and in the code.

File: widget-store/public/app.html [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+              </button>
+            </div>
+          </div>
+        </div>
+      </div>
+    </div>
+
+    <div
+      class="md:hidden w-full grid grid-cols-3 font-semibold font-mono bg-neutral-200"
+    >
```
[the remark this answers is not in the document]
Comment (lanice) [reply widget-store/public/app.html:546]: Done. Hadn't realized that it was not centered on Chrome before.

File: widget-store/migrations/001_initial_schema.js (hunk 4 above)
Comment (lanice) [reply widget-store/migrations/001_initial_schema.js:16]: I updated the README with a description of what it does, and what happens on crash and after recovery. Does the update make sense?

File: widget-store/src/utilities.ts [hunk 6]
```diff
@@ -115,12 +113,34 @@ export class ShopUtilities {
...
+
+    const order = orders[0];
+    if (order.progress_remaining > 1) {
+      await ctxt.client<Order>('orders').where({ order_id: order_id }).update({ progress_remaining: order.progress_remaining - 1 });
+    } else {
+      await ctxt.client<Order>('orders').where({ order_id: order_id }).update({
+        order_status: OrderStatus.DISPATCHED,
+        progress_remaining: 0
+      });
+    }
```
Comment (kraftp) [inline widget-store/src/utilities.ts:143]: Can you check if `nightlyReport`, `getDailySales`, `sendStatusEmail`, and `SalesSummary` are doing anything?  If not, remove them.

**Review by kraftp — APPROVED** [review]
Looks much better!

File: widget-store/src/utilities.ts (hunk 6 above)
Comment (lanice) [reply widget-store/src/utilities.ts:143]: They do things as per the [readme](https://github.com/dbos-inc/dbos-demo-apps/tree/main/widget-store#nightly-sales-reports-optional), i.e. send out an email report IF you configure it. I did not touch those parts. Do we want to keep it, or remove?

Comment (kraftp) [reply widget-store/src/utilities.ts:143]: Let's leave it then

**Review by qianl15 — APPROVED** [review]

**Pull request merged**